Improve handling of "dropped" chunks

erimatnor opened the pull request:
The internal chunk API is updated to avoid returning `Chunk` objects
that are marked `dropped=true` along with some refactoring, hardening,
and cleanup of the internal chunk APIs. In particular, chunks could
previously be returned in a partial state (without all fields set,
partial constraints, etc.) and this is no longer allowed. Further,
lock handling was unclear when joining chunk metadata from different
catalog tables. This is more clear now by having chunks built within
the scan loop so that proper locks are held when reading other
metadata (such as constraints).

This change also fixes issues with dropped chunks that caused chunk
metadata to be processed many times instead of just once, leading to
potential bugs or bad performance.

In particular, since the introduction of the “dropped” flag, chunk
metadata can exist in two states: 1. `dropped=false`
2. `dropped=true`. When dropping chunks (e.g., via `drop_chunks`,
`DROP TABLE <chunk>`, or `DROP TABLE <hypertable>`) there are also two
modes of dropping: 1. DELETE row and 2. UPDATE row and SET
dropped=true.

The deletion mode and the current state of chunk lead to a
cross-product resulting in 4 cases when dropping/deleting a chunk:

1. DELETE row when dropped=false
2. DELETE row when dropped=true
3. UPDATE row when dropped=false
4. UPDATE row when dropped=true

Unfortunately, the code didn't distinguish between these cases. In
particular, case (4) should not be able to happen, but since it did it
lead to a recursing loop where an UPDATE created a new tuple that then
is recursed to in the same loop, and so on.

To fix this recursing loop and make the code for dropping chunks less
error prone, a number of assertions have been added, including some
new light-weight scan functions to access chunk information without
building a full-blown chunk.

erimatnor commented on src/chunk.c:
```diff
@@ -177,9 +177,15 @@ chunk_fill(Chunk *chunk, HeapTuple tuple, TupleDesc desc)
+											get_namespace_oid(chunk->fd.schema_name.data, true));
+		chunk->hypertable_relid = ts_inheritance_parent_relid(chunk->table_id);
+
+		Assert(OidIsValid(chunk->table_id));
```
FWIW, this assert exists in multinode and was hit because we chunks that are marked dropped=true obviously have no real data tables. But this wasn't checked/handled in master.

erimatnor replied:
I couldn't agree more, but that's a major refactoring (which I guess should have been done when this feature was introduced). 

At this point, however, I wanted to do minimal changes to continue with our rebase of the multinode branch. More refactoring, while worthwhile, will make the rebase harder. I could do it if we agree this is important (and I think it needs to happen at some point because the handling of dropped chunk is quite broken at the moment).

gayyappan replied:
Yes, agree that refactoring can be done after code rebase.

gayyappan commented on src/chunk.c:
```diff
@@ -1845,17 +1968,29 @@ ts_chunk_exists_relid(Oid relid)
 	chunk_formdata_fill(&form, ti->tuple, ti->desc);
 
+	if (preserve_chunk_catalog_row && form.dropped)
+		return CHUNK_ALREADY_MARKED_DROPPED;
```
why is this not an error condition? How do we run into this case where we try to mark the same chunk as dropped more than once?

erimatnor replied:
The metadata table is scanned using the UNIQUE <schema, name> index, which doesn't include the dropped column. Therefore, there is no filter on dropped chunks and this acts as the filter instead. 

Also not that there is a case when we'd like to DELETE also chunk entries that are "dropped", so we cannot always filter rows with dropped=true.

gayyappan commented on src/chunk_constraint.c:
```diff
@@ -651,17 +651,23 @@ chunk_constraint_delete_metadata(TupleInfo *ti)
-		get_relation_constraint_oid(chunk->table_id, NameStr(*DatumGetName(constrname)), true));
+	/* Get the chunk relid. Note that, at this point, the chunk table can be
+	 * deleted already */
+	Oid chunk_relid = ts_chunk_get_relid(chunk_id, true);
```
I don't understand the reason for the code change in this file : use ts_chunk_get_relid versus ts_chunk_get_by_id. Could you elaborate why the code is modified here?  It looks like ts_chunk_get_by_id will throw an error if we try to do this on a  chunk that is marked dropped. Isn't that the desired behavior?

erimatnor replied:
There are a couple of reasons. The main one is that I added some Asserts for hardening that is now hit without this function. For example, there's now and Assert that checks that a Chunk cannot have an invalid `table_id`, i.e., you cannot have a Chunk object without the table actually existing. When you DROP a hypertable and it cascades to chunks, the data table is actually removed before we remove the metadata, so this assert would be hit. Thus, we'd need a way to read chunk metadata rows without building complete chunks.

`ts_chunk_get_by_id` allocates a "complete" chunk, and joins in other metadata like constraints, relids for the data tables, etc. When dropping a chunk, this function is called and in various places over-and-over again when the drop cascades to other metadata, like `chunk_constraint_delete_metadata` and `chunk_constraint_drop_constraint`. In contrast, `ts_chunk_get_relid` only reads the minimal data

Ideally, we'd never be able to build `Chunk` objects when they don't really exists (e.g., we should never have a `Chunk` object that has `dropped=true`) but this requires a major refactoring of the code.
This is a lot of unnecessary work that's not needed in many places. In contrast,  `ts_chunk_get_relid` translates the internal ID to a relid by reading _only_ the chunk row metadata.

WireBaron commented on src/chunk.c:
```diff
@@ -1831,12 +1840,126 @@ ts_chunk_get_by_id(int32 id, int16 num_constraints, bool fail_if_not_found)
+
+	found = (num_found == 1 && !form->dropped);
+
+	Assert(num_found == 0 || found);
```
It looks like we'll always hit this assertion if there's a dropped chunk, is that the intention?

erimatnor replied:
No, I guess not. Let me take another look.

erimatnor commented:
@gayyappan @WireBaron I updated the PR with further refactoring and assertions based on feedback. Now it is asserted that every public function in `chunk.{c,h}` that returns `Chunk` objects actually returns "valid" chunks (where, among other things, dropped=false).

erimatnor commented on src/chunk.c:
```diff
@@ -1181,17 +1275,26 @@ ts_chunk_find(Hyperspace *hs, Point *p, bool include_chunks_marked_as_dropped)
+ts_chunk_find(Hypertable *ht, Point *p)
+{
+	Chunk *chunk = chunk_find(ht, p, false);
+	ASSERT_IS_NULL_OR_VALID_CHUNK(chunk);
```
I don't think so since it shouldn't happen. A chunk should be valid or not found.

erimatnor commented on src/chunk.c:
```diff
@@ -739,7 +770,7 @@ chunk_create_postgres_objects_after_lock(Hypertable *ht, Chunk *chunk)
 								chunk->fd.hypertable_id);
 
-	ts_trigger_create_all_on_chunk(ht, chunk);
+	ts_trigger_create_all_on_chunk(chunk, ht->space->main_table_relid);
```
I guess it is actually sufficient (and better) for `ts_trigger_create_all_on_chunk` to only take the `Chunk`

gayyappan replied:
yes, don't need to pass hypertable_relid.

erimatnor commented on src/chunk.c:
```diff
@@ -1123,6 +1155,58 @@ chunk_scan_ctx_get_chunk(ChunkScanCtx *ctx)
+	Assert(count == 0 || count == 1);
+
+	if (count == 0)
+		elog(ERROR, "cannot resurrect a chunk without a tombstone");
```
I realized this probably shouldn't be an error at all. I guess you must either succeed to resurrect a chunk if you already found a stub (i.e., there existed metadata). If you fail, then the only reason would be that someone removed the metadata before you had a chance to resurrect. And, in that case the resurrection should not succeed.

I updated the code to reflect this (also ensuring that resurrection happens while holding the lock on the chunk metadata row).

gayyappan commented on src/chunk.c:
```diff
@@ -1123,6 +1153,62 @@ chunk_scan_ctx_get_chunk(ChunkScanCtx *ctx)
+		 * objects while we hold lock on the metadata row */
+		chunk->constraints = stub->constraints;
+		chunk->cube = stub->cube;
+		chunk_create_table_after_lock(ht, chunk);
```
where is this functionality used - i.e. when do we need to recreate the table versus just using the metadata for a tombstoned chunk?

erimatnor replied:
There are two cases when we create tables: 1. When the chunk is first created (no tombstone) and 2. when it is resurrected (tombstone exists).

gayyappan commented on src/chunk.c:
```diff
@@ -1123,6 +1153,62 @@ chunk_scan_ctx_get_chunk(ChunkScanCtx *ctx)
+	ts_scan_iterator_close(&iterator);
+
+	Assert(count == 0 || count == 1);
+
```
this functions is called when we see the state as dropped. Something is wrong if we try to reconstruct the metadata and it has been deleted from under us. This should error out and not silently ignore.
I am not clear about the changes to this function. Should discuss this further.

erimatnor replied:
Can you explain why it is wrong?

This is my reasoning: with read committed isolation, the current transaction would be able to see changes that another transaction has made as soon as the other transaction is committed. If that other transaction deleted the chunk metadata (e.g., because the continuous aggregate was dropped), then the chunk metadata would be gone by the time we scan it in this function unless we held a lock on the chunk row from the point we created the ChunkStub that is an input parameter to this function.

In other words, how can we guarantee that the the metadata we used to create the ChunkStub remains until we scan it again in this function?

We actually have similar logic when creating the chunk initially to protect against a case when another transaction races to create the chunk before us. We first do a cheap lookup for an existing chunk without grabbing a heavy lock, then if the chunk is not found, we grab the heavy lock (on hypertable) followed by another lookup to ensure that the chunk hasn't been created by someone else before we grabbed the lock.

Thus, the logic is similar here: when we have determined we need to resurrect a chunk, we need to grab a heavy lock (RowExclusiveLock) on the row, but if the row is not found it must mean someone else deleted it before we got the lock. Or, am I missing something?

Note that we still hold the hypertable lock here, but I don't think it protects us here since the path to delete is different from the path to create.

erimatnor replied:
@gayyappan is this resolved now?

gayyappan replied:
yes.

WireBaron commented on src/chunk.c:
```diff
@@ -884,27 +905,31 @@ chunk_fill_from_stub(Chunk *result, ChunkStub *chunk_stub, bool tuplock)
+
+	if (NULL == chunk->cube)
+		chunk->cube = ts_hypercube_from_constraints(chunk->constraints, CurrentMemoryContext);
 	else
```
I know that we don't typically use braces for single lines of code, but given the comment and I think they would improve readability here.  At the very least can you remove the empty line.

WireBaron commented on src/chunk.c:
```diff
@@ -836,39 +842,54 @@ ts_chunk_create_base(int32 id, int16 num_constraints)
 		.nkeys = 1,
 		.scankey = scankey,
 		.data = result,
+		.filter = chunk_tuple_dropped_filter,
```
How does this filter work?  Looking at the function, I'd expect that this would result in the scan not returning a dropped chunk.  However, you also added code below to handle the case where a dropped chunk will return NULL instead of hitting an error.

erimatnor replied:
Dropped chunks are not returned, but the is_dropped boolean is set in the ChunkStubScanCtx to indicate to the caller that there was metadata marked as dropped.

Not sure this answers your question?

k-rus commented on src/chunk.c:
```diff
@@ -1790,53 +1977,195 @@ ts_chunk_get_by_relid(Oid relid, int16 num_constraints, bool fail_if_not_found)
+	if (!found && !missing_ok)
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("chunk with relid %u not found", relid)));
```
What's about when `relid` is not valid? In such case the error can be improved.

erimatnor replied:
This is a rare case and it would actually say that relid is 0. IMO, it is not worth adding a custom error for this.

k-rus replied:
You add the error, so I don't understand why it should not print a meaningful message. I think it is fine to keep the same error code.

erimatnor replied:
I guess I misunderstood what you are asking. Can you explain why the current error is not meaningful and how you'd like it to be changed?

k-rus replied:
In my last comment I was thinking about the other case, where you applied the changes already. Sorry it is my fault.

Actually your first reply answers my question.

k-rus commented on src/chunk.c:
```diff
@@ -1790,53 +1977,195 @@ ts_chunk_get_by_relid(Oid relid, int16 num_constraints, bool fail_if_not_found)
+	if (!OidIsValid(relid) && !missing_ok)
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_SCHEMA),
+				 errmsg("chunk with relid %u not found", relid)));
```
Why is the error given in terms of `relid`, not `chunk_id? Should it be:
```suggestion
				 errmsg("chunk with id %u not found", chunk_id)));
```

k-rus commented on src/chunk.h:
```diff
@@ -101,31 +101,31 @@ typedef enum CascadeToMaterializationOption
 extern void ts_chunk_recreate_all_constraints_for_dimension(Hyperspace *hs, int32 dimension_id);
-extern TSDLLEXPORT void ts_chunk_drop_fks(Chunk *chunk);
-extern TSDLLEXPORT void ts_chunk_create_fks(Chunk *chunk);
+extern TSDLLEXPORT void ts_chunk_drop_fks(Chunk *const chunk);
```
You expose an implementation detail, which only affects the usage of the variable inside the function implementation, and has no sense for the function user. I suggest to remove this and below changes.

erimatnor replied:
The microsoft compiler warns against having different declarations, so if we are to enable warnings as errors, we need to keep declaration and definition consistent. I also think that it is good rule in general to keep consistency.

k-rus replied:
@erimatnor Good to know. What version do you use? Is it through command line or Visual Studio?

erimatnor replied:
I believe I used Visual Studio 2017, both via the command-line and Visual Studio.

k-rus replied:
I haven't noticed these warnings in 2019, but I didn't pay too much attention. I will double check how it works there.

k-rus commented on src/chunk_constraint.c:
```diff
@@ -651,17 +652,23 @@ chunk_constraint_delete_metadata(TupleInfo *ti)
-	 */
-	if (OidIsValid(index_relid))
-		ts_chunk_index_delete(chunk, index_relid, false);
+	if (OidIsValid(chunk_relid))
```
I believe it cannot be the other case, when it is not valid. I suggest to either remove the condition or replace with assert.

erimatnor replied:
It can be invalid, per the comment above.

Basically, there are multiple orders in which events can happen. 

1) Either the chunk and/or constraint is deleted before our own metadata
2) The chunk and/or constraint is deleted after our metadata

This code handles both cases.

k-rus replied:
I understood that the condition is correct, since it specifies `missing_ok = true`, while it was `false` before and will not work correctly with the described case.

k-rus commented on src/chunk_constraint.c:
```diff
@@ -674,15 +681,21 @@ chunk_constraint_drop_constraint(TupleInfo *ti)
 
-	if (OidIsValid(constrobj.objectId))
-		performDeletion(&constrobj, DROP_RESTRICT, 0);
+	if (OidIsValid(chunk_relid))
```
The other condition should not happen. Either remove the condition or replace with assert.

erimatnor replied:
I am not sure I follow. What other condition are you referring to?

k-rus replied:
I mean the `else` case.

k-rus replied:
I.e., `OidIsValid(chunk_relid)`is always true.

k-rus replied:
Actually I am wrong the function call specified `missing_ok = true`.

k-rus replied:
It was `false` before.
I guess the change supports dropped case when no metadata should be deleted and no error should be produced.

k-rus commented:
@erimatnor I forgot to add to my review that the case of deleting dropped chunk is not tested according the code coverage. It is good to have such test.

erimatnor commented:
> @erimatnor I forgot to add to my review that the case of deleting dropped chunk is not tested according the code coverage. It is good to have such test.

I agree. In fact there are many cases with dropping chunks that I think can be better tested. But this PR doesn't change existing functionality so in the interete

erimatnor commented:
Accidentally closed this PR, so reopening.

erimatnor commented on src/chunk.c:
```diff
@@ -836,40 +844,109 @@ ts_chunk_create_base(int32 id, int16 num_constraints)
+	if (NULL == chunkptr)
+		chunkptr = &chunk;
+
+	if (NULL == *chunkptr)
```
The memory for the chunk can be pre-allocated by the caller. For instance, there's one path in the code that allocates on array of chunks (not array of Chunk pointers) and fills them in. In other cases, the memory is generally not pre-allocated.

The questions whether the stub is valid is an orthogonal matter and is checked by `chunk_stub_is_valid`.

erimatnor commented on src/chunk.c:
```diff
@@ -880,31 +957,24 @@ chunk_fill_from_stub(Chunk *result, ChunkStub *chunk_stub, bool tuplock)
-		 * sort them in dimension order.
-		 */
-		ts_hypercube_slice_sort(result->cube);
+	if (num_found != 1)
```
Not sure what is wrong with the current condition given that the only possible value here should be 0 or 1. So, if it is not 1, the chunk is not found.

k-rus replied:
It's confusing due to the assert above. According to other places with similar condition, `num_found == 0` gives one error message, while `num_found > 1` gives another error message.

erimatnor replied:
Sorry, but I am not following how this is confusing. It is asserted that `num_found` can be either 1 or 0, so I don't understand why anything greater than 1 is important here?

In any case, I can change, but I think what you propose because the only viable value one returning from this function below is num_count==1, which is exactly what this error condition is saying (even without asserts).

k-rus replied:
The error message is only suitable for `num_found == 0`. Thus I find it confusing. In any case I don't find it is so important, since the result of execution is correct.

erimatnor commented on src/chunk.c:
```diff
@@ -1406,13 +1563,36 @@ chunk_find_all(Hyperspace *hs, List *dimension_vecs, on_chunk_stub_func on_chunk
-	return chunk_find_all(hs, dimension_vecs, append_chunk, lockmode, num_chunks);
+	Chunk **chunks = chunk_find_all(hs, dimension_vecs, append_chunk, lockmode, num_chunks);
+
+#ifdef USE_ASSERT_CHECKING
```
AFAIK, it depends on how you configured your PostgreSQL build (`--enable-cassert`), which is true for all our `Assert`s. Unless I am mistaken, CodeCov indicates that this path is executed. Are you seeing something else?

k-rus replied:
I see in CodeCov that it doesn't show as executed, but it doesn't show opposite either. I confirm that it should be configured for majority of the tests. So I believe I am wrong. It would be nice to understand why it is not shown in CodeCov, but it is not part of this issue.

erimatnor commented:
@k-rus I looked into the testing of deleting dropped chunks. It turns out that CodeCov indicates that the path is hit when deleting chunks by <schema,name>, but not by <hypertable_id>. I actually refactored the code to handle these two delete cases with common code, since they shared a lot of commonality and now I believe this one case should be hit.

erimatnor commented:
@k-rus I looked into the testing of deleting dropped chunks. It turns out that CodeCov indicates that the path is hit when deleting chunks by <schema,name>, but not by <hypertable_id>. I actually refactored the code to handle these two delete cases with common code, since they shared a lot of commonality and now I believe this one case should be hit.

erimatnor commented on src/chunk.c:
```diff
@@ -612,24 +640,27 @@ get_am_name_for_rel(Oid relid)
  */
 Oid
-ts_chunk_create_table(Chunk *chunk, Hypertable *ht, char *tablespacename)
+ts_chunk_create_table(Chunk *chunk, Hypertable *ht, const char *tablespacename)
```
Unfortunately, this doesn't work because none of the PostgreSQL functions or data types use const. We'd have to do a copy or cast in that case (for instance, the `makeRangeVar` below doesn't accept const).

k-rus replied:
I missed that there were calls to PG functions in this function. I noticed in other similar place and didn't comment there.
